File: BankKataJava/src/test/java/bankingkata/AccountTest.java
package bankingkata;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.mockito.Mockito.*;

class AccountTest {

    private Account account;
    private final TransactionLog transactionLog = mock(TransactionLog.class);
    private final StatementPrinter statementPrinter = mock(StatementPrinter.class);

    @BeforeEach
    void setUp() {
        account = new Account(transactionLog, statementPrinter);
    }

    @Test
    void deposit() {
        account.deposit(500);

        verify(transactionLog).addDeposit(500);
    }

    @Test
    void withdraw() {
        account.withdraw(200);

        verify(transactionLog).addWithdraw(200);
    }

    @Test
    void printStatement() {
        List<Transaction> transactions = List.of(new Transaction());
        when(transactionLog.getAllTransactions()).thenReturn(transactions);

        account.printStatement();

        verify(statementPrinter).print(transactions);

    }
}

File: BankKataKotlin/src/test/kotlin/bankingkata/AccountTest.kt
package bankingkata

import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.verify
import org.junit.jupiter.api.Test

import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.extension.ExtendWith

@ExtendWith(MockKExtension::class)
internal class AccountTest {

    @MockK(relaxUnitFun = true)
    private lateinit var statementPrinter: StatementPrinter

    @MockK(relaxUnitFun = true)
    private lateinit var transcationLog: TransactionLog
    private lateinit var account: Account

    @BeforeEach
    internal fun setUp() {
        account = Account(transcationLog, statementPrinter)
    }

    @Test
    internal fun deposit() {
        account.deposit(200)

        verify {
            transcationLog.addDeposit(200)
        }
    }

    @Test
    internal fun withdraw() {
        account.withdraw(300)

        verify {
            transcationLog.addWithdraw(300)
        }
    }

    @Test
    internal fun printStatement() {
        val transactions = listOf(Transaction())
        every {
            transcationLog.allTransactions
        } returns transactions
        account.printStatement()

        verify {
            statementPrinter.printAll(transactions)
        }
    }
}
